# Homework 02 — White Dwarf Pulsations: Finite Differences

## Problem

DAV (ZZ Ceti) white dwarfs pulsate in several non-radial g-modes with typical periods between 100 and 1500 seconds (1.7–25 minutes). Their flux can be modeled as

$$
F(t)
=
100
+
15\sin\left(\frac{\pi t}{6}\right)
+
3\sin\left(\frac{\pi t}{2}\right),
$$

where $t$ is measured in minutes.

For the finite difference calculations, use

$$
h=\Delta t=1\ \text{minute}.
$$

## Tasks

**a)** Evaluate $F(t)$ at

$$
t=10,\ 11,\ 12,\ 13,\ 14\ \text{minutes}.
$$

In [3]:
import numpy as np

def F(t):
  return 100 + 15*np.sin((np.pi*t)/6) + 3*np.sin((np.pi*t)/2)

#Punto a
tiempos = [10,11,12,13,14]

for t in tiempos:
  print(f"El flujo en t = {t} minutos es: {F(t)}")

El flujo en t = 10 minutos es: 87.00961894323342
El flujo en t = 11 minutos es: 89.5
El flujo en t = 12 minutos es: 100.0
El flujo en t = 13 minutos es: 110.5
El flujo en t = 14 minutos es: 112.99038105676658


**b)** Approximate $F'(12)$ using the **forward**, **backward**, and **centered finite difference** formulas with $h=1$.

In [4]:
#Punto b
h = 1
t_0 = 12

def derivada_forward(f,t,h):
  return (f(t+h)-f(t))/h

def derivada_backward(f,t,h):
  return (f(t)-f(t-h))/h

def derivada_centrada(f,t,h):
  return (f(t+h)-f(t-h))/(2*h)

print(f"La derivada con diferencias finitas hacia adelante en t = 12 es: {derivada_forward(F,t_0,h)}")
print(f"La derivada con diferencias finitas hacia atras en t = 12 es: {derivada_backward(F,t_0,h)}")
print(f"La derivada con diferencias finitas centradas en t = 12 es: {derivada_centrada(F,t_0,h)}")

La derivada con diferencias finitas hacia adelante en t = 12 es: 10.5
La derivada con diferencias finitas hacia atras en t = 12 es: 10.5
La derivada con diferencias finitas centradas en t = 12 es: 10.5


**c)** Calculate the analytical derivative $F'(t)$ and evaluate it at $t=12$. Then calculate the percentage relative error of each finite difference approximation with respect to the analytical derivative.

In [6]:
#Punto c
def derivada_analitica(t):
  return (15*np.pi/6)*np.cos((np.pi*t)/6) + (3*np.pi/2)*np.cos((np.pi*t)/2)

derivada_exacta_12 = derivada_analitica(t_0)

print(f"La derivada exacta en t = 12 es: {derivada_exacta_12}")

error_forward = abs((derivada_exacta_12-derivada_forward(F,t_0,h))/derivada_exacta_12)*100
error_backward = abs((derivada_exacta_12-derivada_backward(F,t_0,h))/derivada_exacta_12)*100
error_centrada = abs((derivada_exacta_12-derivada_centrada(F,t_0,h))/derivada_exacta_12)*100

print(f"Error relativo hacia adelante: {error_forward}%")
print(f"Error relativo hacia atras: {error_backward}%")
print(f"Error relativo centrada: {error_centrada}%")

La derivada exacta en t = 12 es: 12.566370614359172
Error relativo hacia adelante: 16.443654876754945%
Error relativo hacia atras: 16.443654876754945%
Error relativo centrada: 16.443654876754945%


## Conclusion

For $h=1$ minute, the forward, backward, and centered finite difference methods produce the same approximation,

$$
F'(12)\approx 10.5,
$$

with a relative error of approximately $16.44\%$ with respect to the analytical derivative.

Although the centered difference has a truncation error of order $O(h^2)$, while the forward and backward differences have errors of order $O(h)$, this does not guarantee a smaller error for every particular value of $h$.